rfcs/0002-mvp.md

```diff
@@ -0,0 +1,159 @@
+supa start    # Starts Supabase (by running docker-compose), and pulls up the database
+supa eject    # See the .supabase/docker-compose.yml details below
+```
+This will likely need: 
+
+- `supabase.js` - a config file with the Supabase Project details
+- `.supabase/config.txt` - CLI config and details
+- `.supabase/docker-compose.yml` - for running Supabase locally. We may not even need this, as we can just run it directly from the `npm_modules` folder, which would prevent conflicts when we add upgrades or make changes. If we do generate a file, this should be stored in the `.supabase` folder so that it doesn't conflict with the user's docker-compose.yml file. When the user runs `supa eject`, it can move the file out to their root folder (similar to a `cra eject` model), and then the user can manage this themselves.
+
+
+For the database scaffold, [this](https://gitlab.com/delibrium/delibrium-postgrest/-/tree/master/database) format is very nice, although I think we can make it a bit clearer. Here is a suggested format:
+
+```sh
+│   # Data: database fixed/seed data.
+├── data
+│   ├── some_data.{sql|csv|json}        
+│   # Models: Tables and Views
+├── models
+│   ├── public.table                    # Personal preference to explicitly add the schema, and we should prefer dot notation over nesting.
+│   │   ├── data.{sql|csv|json}         # For convenience, table specific data can be stored here too.   
+│   │   ├── model.sql                   # This is the core. We should also figure out a nice, readable format for SQL files. Should include constraints and indexes and table/column comments.
+│   │   ├── permissions.sql             # Grants and Policies.
+│   │   ├── test.sql                    # PGTAP tests?
+│   │   └── triggers.sql                # table triggers 
+│   # Functions/RPC: functions. At the moment the function are written in SQL. If we want to support something like plv8, we would have to include a builder (like webpack) to bundle the function.
+├── functions                                
+│   ├── public.func1.sql                
+│   ├── public.func2.sql│   
+│   # Postgres roles and grants.
+├── roles                                
+│   ├── role_name.sql    
+│   # Types: custom database types
+└── types
+    ├── _init.sql
+    ├── public.some_type.sql
+    └── extensions.some_type.sql
+```
+
+
+### Branching
+
+This is going to be a key part of Supabase development. Since so many of our users are on Vercel/Netlify, we should have the ability to support [branch/preview-development](https://vercel.com/docs/platform/deployments).
+
+We can do this by connecting to a different database, an idea that I got from [pgsh](https://github.com/sastraxi/pgsh). For example:
+
+- `postgresql://postgres@localhost:5432/master`
+- `postgresql://postgres@localhost:5432/develop`
+- `postgresql://postgres@localhost:5432/chore_add_types`
```

Comment:
On these different envs, would the users also get different realtime/postgrest instances?

For preview workflows, the users would like to interact with the full stack right?

Reply:
This completely slipped my mind. Yes they are going to need it. Hmm, how will we keep the costs down? 

Perhaps it will have to be "local only" for now? It should be manage-able with the docker files

Reply:
> Hmm, how will we keep the costs down?

A temporary postgrest with `db-pool = 1`(one connection max) should be lightweight, it would fit on the same instance. Realtime would need to limit its pool as well, but we'd definitely need to set its systemd `MemoryLimit` to be really low. But I'm not sure how low it can get.

> Perhaps it will have to be "local only" for now? It should be manage-able with the docker files

Yes, local-only it's a good start. Other envs can be done in a later RFC - they need more thought.

Reply:
We can also take advantage of systemd [socket activation](https://blog.developer.atlassian.com/docker-systemd-socket-activation/), which essentially lets a service sleep until a request comes. That coupled with a systemd timer to stop the service after some idle time([so question](https://unix.stackexchange.com/questions/470880/deactivate-a-systemd-service-after-idle-time)), would make resource usage minimum.